internal/terraform/node_resource_validate.go

```diff
@@ -478,14 +439,75 @@ func (n *NodeValidatableResource) validateResource(ctx EvalContext) tfdiags.Diag
 	return diags
 }
 
-func validateCheckRules(ctx EvalContext, crs []*configs.CheckRule, self addrs.Referenceable) tfdiags.Diagnostics {
+func (n *NodeValidatableResource) evaluateExpr(ctx EvalContext, expr hcl.Expression, wantTy cty.Type, self addrs.Referenceable, keyData instances.RepetitionData) (cty.Value, tfdiags.Diagnostics) {
+	var diags tfdiags.Diagnostics
+
+	refs, refDiags := lang.ReferencesInExpr(expr)
+	diags = diags.Append(refDiags)
+
+	scope := ctx.EvaluationScope(self, keyData)
```

Comment:
I like this idea of a resource-node-specific expression evaluate function, but it makes me wonder about having this function itself call into `stubRepetitionData` to decide `keyData` and `self`, since it seems like that would make the callers a little simpler and it would be symmetrical with your simplification of `n.evaluateBlock`.

(I guess there would still be the need to distinguish between `self` being set or not depending on the situation, so probably that's why you structured it this way? There is no such need for `evaluateBlock` because it's only for provisioners today and they always have access to `self`.)

This is a style/organization thing rather than a behavior thing so I don't intend it as a blocker for moving forward.

Reply:
Yes, the distinction between `self` being available or not was the reason for this awkward signature. Given the need for that context to be fed in, the only alternative I could see was something like:

```go
func (n *NodeValidatableResource) evaluateExpr(
    ctx EvalContext,
    expr hcl.Expression,
    wantTy cty.Type,
    canReferToSelf bool) (cty.Value, tfdiags.Diagnostics) {
```

…which didn't seem better to me than passing the self & keydata values.